freedomofpress/securedrop
# Fixes reboot-and-wait-for strategy across all platforms

## Status

Ready for review.

## Description of Changes

Fixes #951. Changes introduced in #1666 caused variability in the "reboot and wait for" strategy (RAWFS) used in several places throughout the site config. To merge #1666 and boost the automated tests, we opted for the RAWFS that worked well with AWS EC2 instances—but failed on Vagrant  local VMs. The new strategy works equally well in both contexts.

Changes proposed in this pull request:

* Notifies a single handler for reboot-and-wait-for (DRY)
* Reduces number of reboots from 3 to 1 on clean staging install

There are some non-obvious side-effects of the handler approach, namely:

  1. The handler is declared in the `common` role and therefore is only available to roles within the same play. 
  2. The only way to notify a handler is to have a task that reports changed=true; thus the dummy tasks that simply run `/bin/true` based on file existence.

Down the road I'd like to move toward leaning more heavily on upstream for the Ansible config, which would allow us to remove many of the custom roles for common hardening actions, e.g. sysctl and iptables management. In particular, we have a [well-maintained grsecurity role](https://github.com/freedomofpress/ansible-role-grsecurity) that we're _not_ using in this project yet. Once we include that work, we'll be in a much stronger position. 

## Testing

  1. Confirm CircleCI staging env passes.
  2. Confirm a fresh staging build on Vagrant VMs passes.

## Deployment

The changes presented here modify the implementation of reboots in the playbooks, but with an eye toward preserving existing functionality. That said, vigorous QA on hardware is warranted, although I suggest we include that on the general 0.4 QA testing list.

## Checklist

### If you made changes to the app code:


### If you made changes to the system configuration:

**Note:** Had to bump Ansible down to v2.2.1, since 2.2.2 showed errors with testinfra. If you receive the same results, say so, and we'll make sure to address in #1146.

- [ ] Testinfra tests pass on the mon-staging VM
